### ISSUE TITLE
Implementation Request: Layer Normalization

### ISSUE THREAD

**Issue by juesato** [Element-Research/rnn#323]
I've heard promising things about [layer normalization[(http://arxiv.org/pdf/1607.06450.pdf).

@nicholas-leonard Do you think this is worth including? Arguably it should be inside the `nn` package since it can in theory be used for non-recurrent networks as well, but my impression is that it's very useful for recurrent architectures, and not very useful for non-recurrent architectures.

**Comment by iikulikov** [comment]
hi @juesato , the thing is that original paper does not specify exactly how to apply Layer Normalization for LSTM node, in a very "deep" case, it could be on the activation after the gates,  from my experiments with BLSTM-HMM acoustic modeling for ASR this LLN didnt help at all, though I lost the possibility to use fast CUDA lstm kernel.

In comparison to that, simple MLP together with layer normalization gave a huge speed up during training  even in comparison to Batch Normalization.

If you will implement it in the scope of this rnn package, please provide your improvement here, thanks!

**Comment by juesato** [comment]
@iikulikov Sorry for the slow response, just revisiting this. I'm not sure which part is unclear, the paper seems fairly explicit in the appendix that they used layer normalization on the cell state for the MNIST/DRAW experiments, and in the teaching machines to read and comprehend experiment, they used layer normalization on the cell state and separately on the input and previous hidden state.

FWIW, I've implemented layer normalization on the PTB language model example and it seems to hurt results. Out of curiosity, what was the experiment with the MLP which was helped by layer normalization?

**Comment by iikulikov** [comment]
@juesato yes, exactly for LN only hurts the lstm model for phoneme classification. But, for general MLP classification task with notMnist - it gave really nice speed improvement. MLP was with 2 layers with 2000 hidden units in each layer.

**Comment by juesato** [comment]
@iikulikov I'm surprised layer normalization has any noticeable benefit on such a shallow network. Do you still have the training plots from those experiments?

**Comment by nicholas-leonard** [comment]
@juesato The LN-LSTM seems to have great potential. The validation error convergence is fast. Can we build this as a subclass or method of `SeqLSTM`? Like `LNSeqLSTM` or `lstm:layernormalize()`.

**Comment by nicholas-leonard** [comment]
@iikulikov I haven't read the paper in detail to say if it reproducible or not. Worse case, we can contact the authors for more implementation details. They almost always answer when you mention reproducibility, and when they have time. They are often very busy, but they want you to help them validate their experiments.

**Comment by amartya18x** [comment]
I did try the Layer Normalization in some code of mine with torch-autograd. It does give a faster learning as claimed in the paper in my experiments. I think it is a good think to implement in the SeqLSTM class.
